### PR TITLE
VueUI Titlebars, Resizing, and Scrolling

### PR DESCRIPTION
**Pull request by JohnWildkins**
TL;DR: Goodbye gross white BYOND titlebar ruining my immersion, hello beautiful theme-matched titlebars of the future. Screenshots below.

- Removes the BYOND titlebar (and thus resizing / dragging features) and replaces them with Vue/JS replacements, which may be styled as needed.
- Added a debug button that appears when ~~UIDEBUG is defined~~ if the user has developer permissions, thereby allowing UI developers to toggle debug mode on/off at will in-game
- A few minor styling tweaks (like adding a parameter to have buttons be an icon only so they don't have an extra margin-right sticking out)

Pretty much all of the JS heavy lifting here was ported from /tg/ and then butchered by yours truly, to get it working with our Vue system rather than their React-based system. There are no doubt more than a few JS warcrimes in here, so I'd appreciate a thorough, but gentle, review of things, as well as a decent bit of testing.

### CODE REVIEW

**Comment by JohnWildkins** [comment]
!review

File: vueui/src/assets/icons.scss [hunk 1]
```diff
@@ -267,4 +267,4 @@ $fa-css-prefix: "ic";
             }
         }
     }
-}
\ No newline at end of file
+}
```
Comment (Amunak) [inline vueui/src/assets/icons.scss:270]: You did not change this file otherwise, so this change really shouldn't be in the commit.

File: vueui/src/assets/theme-nano.scss [hunk 2]
```diff
@@ -130,22 +162,32 @@
...
+        top: -1px;
+        right: 0;
+        width: 45px;
+        height: 32px;
+        font-size: 20px;
+        line-height: 31px;
+        text-align: center;
+
+        color: color.change($text-color, $alpha: 0.5);
+        background-color: $header-bg;
```
Comment (Amunak) [inline vueui/src/assets/theme-nano.scss:183]: The close element is on top of the title wrapper, correct? In that case it shouldn't need background-color set, just keep it transparent and we can even have images in there.

Additionally, clickable elements like this should probably have `cursor: pointer` set.

File: vueui/src/collections.js [hunk 3]
```diff
@@ -0,0 +1,271 @@
+/**
+ * @file
+ * @copyright 2020 Aleksej Komarov
+ * @license MIT
+ */
+
```
Comment (Amunak) [inline vueui/src/collections.js:6]: If I understand it correctly most of the functions here could probably be simplified quite a lot with some modern Object.* calls.

File: vueui/src/components/header/default.vue [hunk 4]
```diff
@@ -1,24 +1,60 @@
...
+  },
+  methods: {
+    closeUI() {
+      Utils.sendToTopic({'vueuiclose': 1});
+    },
+    startDragging($event) {
+      dragStartHandler($event);
+    },
+    activateDebug() {
+      var offset = this.modularComputer ? '72px' : '36px';
```
Comment (Amunak) [inline vueui/src/components/header/default.vue:48]: You should be able to easily calculate this dynamically by getting the height of the title wrapper (check `.offsetHeight` or similar on the element). That would allow you to remove the hardcoded check (and simplify lives of people if we ever have more UI styles than simple and modular computer).

Alternatively the calculation could be delegated to the stylesheet by just flipping a class on `body` or the like.

File: vueui/src/components/header/default.vue [hunk 5]
```diff
@@ -28,20 +64,24 @@ export default {
 .uiTitleWrapper {
   margin: -8px -8px 10px;
   position: relative;
-  padding-bottom: 8px;
   float: none;
   background-color: #b9b9b9;
   clear: both;
+  user-select: none;
+  -ms-user-select: none;
```
Comment (Amunak) [inline vueui/src/components/header/default.vue:71]: Do we not have/use autoprefixer? We should probably fix that.

File: vueui/src/components/header/default.vue [hunk 6]
```diff
@@ -28,20 +64,24 @@ export default {
...
+.uiTitleWrapper > div:nth-child(1) {
+  height: 32px;
 }
 
 .uiTitleText {
   position: relative;
   display: inline-block;
-  top: 4px;
-  left: 18px;
+  left: 20px;
```
Comment (Amunak) [inline vueui/src/components/header/default.vue:82]: I'm not sure I like positioning everything with magic values like this. Is there a particular reason why the elements can't just follow regular flow within the title wrapper, one going after another?

File: vueui/src/components/header/modular-computer.vue [hunk 7]
```diff
@@ -23,6 +23,11 @@
 export default {
   data() {
     return this.$root.$data
+  },
+  mounted: function() {
+    this.$nextTick(function() {
+      document.getElementById('content').style.height = 'calc(100vh - 72px)';
+    })
```
Comment (Amunak) [inline vueui/src/components/header/modular-computer.vue:30]: Same situation as in the other files - this function should be unnecessary if you get the title bar height dynamically.

File: vueui/src/components/view/machinery/vending.vue [hunk 8]
```diff
@@ -188,4 +188,8 @@ p.danger {
   opacity: 1;
 }
 
-</style>
\ No newline at end of file
+#content > div {
+  overflow-x: hidden;
```
Comment (Amunak) [inline vueui/src/components/view/machinery/vending.vue:192]: Any chance that this would "silently hide" content?

File: vueui/src/components/vui/button.vue [hunk 9]
```diff
@@ -1,6 +1,6 @@
 <template>
   <div @click="senddata()" class="button" :disabled="$root.$data.status < 2 || this.disabled">
-    <div v-if="icon" class="uiIcon16" :class="'ic-' + icon"/>
+    <div v-if="icon" class="uiIcon16" :class="[this.iconOnly ? '' : 'mr4', 'ic-' + icon]"/>
```
Comment (Amunak) [inline vueui/src/components/vui/button.vue:3]: You could probably just use the global `mr-1` class or similar.

File: vueui/src/utils.js [hunk 10]
```diff
@@ -30,4 +30,4 @@ export default {
   dotNotationRead(object, key) {
     return key.split('.').reduce((a, b) => a[b], object);
   }
-}
\ No newline at end of file
+}
```
Comment (Amunak) [inline vueui/src/utils.js:33]: Again a format change in a non-touched file x)

**Review by Amunak — COMMENTED** [review]
I've gone through all the code but I also want to test this myself and see how it actually behaves and if any improvements can be made there. So not commenting on the JS yet (though it seems nowhere as terrible as you tried to make it on Discord :P).

Amazing work anyway, I'm eager to have this in-game.

File: code/modules/vueui/ui.dm [hunk 11]
```diff
@@ -133,14 +133,16 @@ main ui datum.
 <!DOCTYPE html>
 <html>
 	<head>
-		<meta http-equiv="X-UA-Compatible" content="IE=edge">
-		<meta charset="UTF-8">
-		<link rel="stylesheet" type="text/css" href="vueui.css">
+		<meta http-equiv="X-UA-Compatible" content="IE=edge"/>
+		<meta charset="UTF-8"/>
+		<meta id="vueui:windowId" content="[windowid]"/>
+		<link rel="stylesheet" type="text/css" href="vueui.css"/>
```
[the remark this answers is not in the document]
Comment (JohnWildkins) [reply code/modules/vueui/ui.dm:139]: XML requests are a bit stingier and require having end slashes on every tag. Without it I was getting malformed XML errors, although technically BYOND itself didn't seem to care.

File: vueui/src/collections.js (hunk 3 above)
Comment (JohnWildkins) [reply vueui/src/collections.js:6]: Probably. I'm not comfortable enough in JS (especially modern JS) to say, though.

File: vueui/src/components/header/default.vue (hunk 4 above)
Comment (JohnWildkins) [reply vueui/src/components/header/default.vue:48]: Yeah, me at 3 am didn't even think of just using classes for this.

File: vueui/src/components/header/default.vue (hunk 5 above)
Comment (JohnWildkins) [reply vueui/src/components/header/default.vue:71]: Probably.

File: vueui/src/components/view/machinery/vending.vue (hunk 8 above)
Comment (JohnWildkins) [reply vueui/src/components/view/machinery/vending.vue:192]: Probably not. I specifically put it in the scoped SCSS of vending machines so that the scrollbar wouldn't appear for the vending UI alone, since the vending UI is `display:flex` and grows with the viewport.

**Comment by JohnWildkins** [comment]
Submitted tweaks based on review, and made it so `#define UIDEBUG` is no longer necessary to debug a UI. The bug icon (and thus, access to debug information) is available to all users with R_DEV (Developer) permissions

**Review by Amunak — APPROVED** [review]

**Comment by JohnWildkins** [comment]
The tendency to break and overall be unresponsible that dragging titlebars had has been resolved thanks to @amunak's help!

**Review by Amunak — APPROVED** [review]

File: code/modules/vueui/ui.dm [hunk 12]
```diff
@@ -124,23 +124,23 @@ main ui datum.
   * @return html code - text
   */
 /datum/vueui/proc/generate_html(var/css_tag)
-#ifdef UIDEBUG
-	var/debugtxt = "<div id=\"dapp\"></div>"
-#else
 	var/debugtxt = ""
-#endif
+	if(user && check_rights(R_DEV, FALSE, user=user))
+		debugtxt = "<div id=\"dapp\"></div>"
```
Comment (Karolis2011) [inline code/modules/vueui/ui.dm:129]: Note: UIDEBUG also makes in each open call resend assets in case if they changed instead of using cached assets in resource file.

File: code/modules/vueui/ui.dm [hunk 13]
```diff
@@ -243,6 +247,12 @@ main ui datum.
   */
 /datum/vueui/Topic(href, href_list)
 	. = update_status(FALSE)
+	if(href_list["vueuihrefjson"])
+		// this has to be up here or you can't close windows you're too far from
+		var/json_href = json_decode(href_list["vueuihrefjson"])
+		if(json_href && json_href["vueuiclose"])
+			close()
+			return
```
Comment (Karolis2011) [inline code/modules/vueui/ui.dm:255]:
```suggestion
	if(href_list["vueuiclose"])
			close()
			return
```

File: vueui/src/assets/global.scss [hunk 14]
```diff
@@ -15,6 +16,41 @@ body {
...
+    &.mcomputer {
+        height: calc(100vh - 72px);
+    }
+    &.uiDebug {
+        height: calc(50vh - 36px);
+        &.mcomputer {
+            height: calc(50vh - 72px);
+        }
+    }
+}
```
Comment (Karolis2011) [inline vueui/src/assets/global.scss:35]: Please refrain from using `calc` as IE badly supports it. See https://caniuse.com/?search=calc

File: vueui/src/collections.js [hunk 15]
```diff
@@ -0,0 +1,271 @@
+/**
+ * @file
+ * @copyright 2020 Aleksej Komarov
+ * @license MIT
```
Comment (Karolis2011) [inline vueui/src/collections.js:4]: License is uncompatible with codebase licence (I think)

File: code/modules/vueui/ui.dm [hunk 16]
```diff
@@ -157,6 +157,9 @@ main ui datum.
 	<script type="application/json" id="initialstate">
 		[generate_data_json()]
 	</script>
+	<script type="text/javascript">
+		window.__windowId__ = document.getElementById('vueui:windowId').getAttribute('content');
+	</script>
```
Comment (Karolis2011) [inline code/modules/vueui/ui.dm:162]: This should be put inside `vueui.js` file

**Review by Karolis2011 — COMMENTED** [review]
It looks fine ish, but some things are really unnecessery. If you do not mind I migth try myself reimplementing some functions in neater way else This has to go on. Also cometed individual lines that should be adressed.

File: vueui/src/collections.js (hunk 15 above)
Comment (skull132) [reply vueui/src/collections.js:4]: Wrong. MIT can be incorporated into AGPL. AGPL cannot be incorporated into MIT.

Comment (NonQueueingMatt) [reply vueui/src/collections.js:4]: I've asked Skull and it seems MIT is fine with our license.

**Comment by Karolis2011** [comment]
Sorry for not doing this before prevous comment, but this PR is far from perfect. To name some issues:

- Drag handle offest is all wonky
- You have no acounted themes. It's all broken for other themes. [image](https://user-images.githubusercontent.com/11229891/99185008-af302600-274f-11eb-97da-fc6522d954c0.png)
- Window closing should be handled simultaneously at client/Topic() instead of vueui/Topic()
- Extra js files by other author is not optimal.
- And some things are done not vuejs way.

File: vueui/src/collections.js [hunk 17]
```diff
@@ -0,0 +1,271 @@
...
+ * iteratee returns truthy for. The predicate is invoked with three
+ * arguments: (value, index|key, collection).
+ *
+ * If collection is 'null' or 'undefined', it will be returned "as is"
+ * without emitting any errors (which can be useful in some cases).
+ *
+ * @returns {any[]}
+ */
+export const filter = iterateeFn => collection => {
+  if (collection === null && collection === undefined) {
```
Comment (skull132) [inline vueui/src/collections.js:76]:
```suggestion
  if (collection === null || collection === undefined) {
```
lmao

File: vueui/src/collections.js [hunk 18]
```diff
@@ -0,0 +1,271 @@
...
+ * thru an iteratee function. The iteratee is invoked with three
+ * arguments: (value, index|key, collection).
+ *
+ * If collection is 'null' or 'undefined', it will be returned "as is"
+ * without emitting any errors (which can be useful in some cases).
+ *
+ * @returns {any[]}
+ */
+export const map = iterateeFn => collection => {
+  if (collection === null && collection === undefined) {
```
Comment (skull132) [inline vueui/src/collections.js:103]:
```suggestion
  if (collection === null || collection === undefined) {
```
el2

**Review by skull132 — CHANGES_REQUESTED** [review]

File: vueui/src/assets/global.scss (hunk 14 above)
Comment (Amunak) [reply vueui/src/assets/global.scss:35]: Shouldn't be an issue in this case, the bad support isn't in simple cases like this thankfully. But I agree we have to be cautious about it.

File: code/modules/vueui/ui.dm (hunk 12 above)
Comment (JohnWildkins) [reply code/modules/vueui/ui.dm:129]: yeah, that's why i didn't remove the UIDEBUG define

**Review by skull132 — APPROVED** [review]
